torch/_inductor/compile_fx.py

```diff
@@ -219,88 +224,630 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
     return static_input_idxs
 
 
-def cudagraphify_impl(model, inputs, static_input_idxs=()):
+@dataclasses.dataclass(frozen=True)
+class GraphID:
+    id: int
+
+
+class CUDAGraphTapeManger(object):
+    """
+    Groups individual recordings or executions of cuda graphs into a tape of multiple cuda graphs
+    and checks required invariants.
+
+    When graphs are recorded in the same tape, it enforces subsequent execution
+    to follow the same order and have the same output tensor livespans. To remove
+    unnecessary coupling of cuda graphs (and additional imposed invariants),
+    the tape manager will end a currently recording tape whenever it is valid  - when
+    the memory pool no longer has any live allocations.
```

Comment:
This is bad for perf no? Once you finish a cuda graph tape, the memory pool associated with the tape is permanently inaccessible. Even if the pool has no more live allocations, if you continue on the tape you can reuse this memory.

As for what the right way to detect if the tape is done or not, I would probably have suggested snipping it when you "revisit" a node that you've already seen before.

Reply:
All of the tapes use the same memory pool. Will add comment to make that clear.

Reply:
I feel like there should be more invariants on the tapes. Can a subgraph be a member of multiple tapes?

Reply:
Not right now. If you run 

`foo -> foo2` that will be a single recording. On the next execution, if you ran `foo -> foo3`, the caching allocator for the memory pool would not represent the current state of allocations, so you not trivially trace `foo3` then with the same memory pool. 

You would need to make the cuda caching allocator resumable by snapshotting its state at the end of the foo recording, resetting it, and then record `foo3`. Was planning on doing this in next pr on the stack (and won't merge until they are all ready).

Reply:
sounds complicated lol